File: 실버 2 (7문제)/11279_최대 힙.py
# 설명 링크 : https://moon-gd.tistory.com/142
import sys


class Heap:
    def __init__(self):
        self.list = [-1, ]
        self.size = 0

    def add(self, number):
        # create : new node
        self.list.append(number)

        # update : size of heap
        self.size += 1

        # initialize : index of new node
        new_index = self.size

        # heapify
        while new_index > 1:
            # new_node's key is bigger than parent node's key
            if self.list[new_index] > self.list[new_index // 2]:
                # exchange each other
                self.list[new_index], self.list[new_index // 2] = self.list[new_index // 2], self.list[new_index]

                # update : new index
                new_index //= 2
            else:
                break

    def delete(self):
        # if empty heap
        if self.size == 0:
            return 0

        # initialize : save root key to return later
        root_key = self.list[1]

        # update : exchange root node and last node
        self.list[1], self.list[self.size] = self.list[self.size], self.list[1]

        # update : reduce size by 1
        self.size -= 1

        # delete : previous root node
        del self.list[-1]

        # initialize : index
        index = 1

        # heapify
        while index <= self.size:
            # left/right child nodes exist
            if index * 2 + 1 <= self.size:
                # break : heapify is done
                if self.list[index] >= self.list[index * 2] and self.list[index] >= self.list[index * 2 + 1]:
                    break
                else:
                    # left child node has the biggest key
                    if self.list[index * 2] >= self.list[index] and self.list[index * 2] >= self.list[index * 2 + 1]:
                        # update : exchange left child and current node
                        self.list[index], self.list[index * 2] = self.list[index * 2], self.list[index]

                        # update : multiply index by 2
                        index *= 2

                    # right child node has the biggest key
                    elif self.list[index * 2 + 1] >= self.list[index] and self.list[index * 2 + 1] >= self.list[index * 2]:
                        # update : exchange right child and current node
                        self.list[index], self.list[index * 2 + 1] = self.list[index * 2 + 1], self.list[index]

                        # update : multiply index by 2 and plus 1
                        index = index * 2 + 1

            # only left child node exists
            elif index * 2 <= self.size:
                # left child node's key is bigger
                if self.list[index * 2] >= self.list[index]:
                    # update : swap child node's key and current node's
                    self.list[index], self.list[index * 2] = self.list[index * 2], self.list[index]

                    # update : multiply index by 2
                    index *= 2

                # current node's key is bigger
                else:
                    break

            # break : no child node exists
            else:
                break

        return root_key


# to get input faster
q_input = sys.stdin.readline

# input : number of operations
N = int(q_input())

# declare : heap
heap = Heap()

for i in range(N):
    # input : operation
    operation = int(q_input())

    # add
    if operation:
        heap.add(operation)
    # delete
    else:
        print(heap.delete())

File: 실버 2 (8문제)/11279_최대 힙.c
// 설명 링크 : https://moon-gd.tistory.com/142
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 100001

// declare : heap array
int heap[MAX];

// initialize : heap size
int heap_size = 0;

// exchange a and b
void swap(int* a, int* b);

// heap functions
void add(int number);
int delete_root();

int main() {
	// declare : number of operations
	int N = 0;

	// input : get number of operations
	scanf("%d", &N);

	// declare : for getting inputs
	int operation;

	for (int i = 1; i <= N; i++) {
		// input : get operation
		scanf("%d", &operation);

		// add to heap
		if (operation) {
			add(operation);
		}
		// delete root node
		else {
			printf("%d\n", delete_root());
		}
	}

	return 0;
}

// exchange a and b
void swap(int* a, int* b) {
	int temp = *a;
	*a = *b;
	*b = temp;
}

void add(int number) {
	// create : add new node to heap
	heap[++heap_size] = number;

	for (int i = heap_size; i > 1; i = i / 2) {
		// child node's key is bigger than root node's
		if (heap[i] > heap[i / 2]) {
			swap(&heap[i], &heap[i / 2]);
		}
		// break : new node find it's proper location
		else {
			break;
		}
	}
}

int delete_root() {
	// return 0 : empty heap
	if (heap_size == 0) {
		return 0;
	}
	// if heap is not empty
	else {
		// initialize : a variable that will be returned
		int root = heap[1];

		// update : swap root node and last node
		swap(&heap[1], &heap[heap_size]);

		// update : heap size
		heap_size--;

		for (int i = 1; i <= heap_size;) {
			// left & right child nodes exist
			if (i * 2 + 1 <= heap_size) {
				if (heap[i] >= heap[i * 2] && heap[i] >= heap[i * 2 + 1]) {
					break;
				}
				else {
					if (heap[i * 2] >= heap[i] && heap[i * 2] >= heap[i * 2 + 1]) {
						// update : swap left node and root node
						swap(&heap[i], &heap[i * 2]);

						// update : i
						i = i * 2;
					}
					else if (heap[i * 2 + 1] >= heap[i] && heap[i * 2 + 1] >= heap[i * 2]) {
						// update : swap right node and root node
						swap(&heap[i], &heap[i * 2 + 1]);

						// update : i
						i = i * 2 + 1;
					}
				}
			}
			// only left child node exists
			else if (i * 2 <= heap_size) {
				// left child's key is bigger than root node's
				if (heap[i * 2] >= heap[i]) {
					// update : swap left node and root node
					swap(&heap[i], &heap[i * 2]);

					// update : i
					i = i * 2;
				}
				else {
					break;
				}
			}
			// no child exists
			else {
				break;
			}
		}

		return root;
	}
}